exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java

```diff
@@ -0,0 +1,342 @@
+     * as the user changes system options. This value captures the value
+     * calculated at the time that this lease was granted.
+     */
+    private long queryMemory;
+
+    public DistributedQueueLease(QueryId queryId, String queueName,
+                    DistributedLease lease, long queryMemory) {
+      this.queryId = queryId;
+      this.queueName = queueName;
+      this.lease = lease;
+      this.queryMemory = queryMemory;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Lease for %s queue to query %s",
+          queueName, QueryIdHelper.getQueryId(queryId));
+    }
+
+    @Override
+    public long queryMemoryPerNode() { return queryMemory; }
+
+    @Override
+    public void release() {
+      DistributedQueryQueue.this.release(this);
+    }
+
+    @Override
+    public String queueName() { return queueName; }
+  }
+
+  /**
+   * Exposes a snapshot of internal state information for use in status
+   * reporting, such as in the UI.
+   */
+
+  @XmlRootElement
+  public static class ZKQueueInfo {
+    public final int smallQueueSize;
+    public final int largeQueueSize;
+    public final double queueThreshold;
+    public final long memoryPerNode;
+    public final long memoryPerSmallQuery;
+    public final long memoryPerLargeQuery;
+
+    public ZKQueueInfo(DistributedQueryQueue queue) {
+      smallQueueSize = queue.configSet.smallQueueSize;
+      largeQueueSize = queue.configSet.largeQueueSize;
+      queueThreshold = queue.configSet.queueThreshold;
+      memoryPerNode = queue.memoryPerNode;
+      memoryPerSmallQuery = queue.memoryPerSmallQuery;
+      memoryPerLargeQuery = queue.memoryPerLargeQuery;
+    }
+  }
+
+  public interface StatusAdapter {
+    boolean inShutDown();
+  }
+
+  /**
+   * Holds runtime configuration options. Allows polling the options
+   * for changes, and easily detecting changes.
+   */
+
+  private static class ConfigSet {
+    private final long queueThreshold;
+    private final long queueTimeout;
+    private final int largeQueueSize;
+    private final int smallQueueSize;
+    private final double largeToSmallRatio;
+    private final double reserveMemoryRatio;
+    private final long minimumOperatorMemory;
+
+    public ConfigSet(SystemOptionManager optionManager) {
+      queueThreshold = optionManager.getOption(ExecConstants.QUEUE_THRESHOLD_SIZE);
+      queueTimeout = optionManager.getOption(ExecConstants.QUEUE_TIMEOUT);
+      largeQueueSize = (int) optionManager.getOption(ExecConstants.LARGE_QUEUE_SIZE);
+      smallQueueSize = (int) optionManager.getOption(ExecConstants.SMALL_QUEUE_SIZE);
+      largeToSmallRatio = optionManager.getOption(ExecConstants.QUEUE_MEMORY_RATIO);
+      reserveMemoryRatio = optionManager.getOption(ExecConstants.QUEUE_MEMORY_RESERVE);
+      minimumOperatorMemory = optionManager.getOption(ExecConstants.MIN_MEMORY_PER_BUFFERED_OP);
+    }
+
+    @Override
+    public boolean equals(Object other) {
```

Comment:
Well, this is disputable. You may leave it as is but I still disagree.

Reply:
Looks like Drill convention is that `equals()` is to be used only for collections (where `hashCode()` is also needed.) To avoid the misuse of `equals()` in this use case, I renamed the method to `isSameAs()` with the argument being of the same type.

I understand the issue, "equality" (or "sameness") is a deep philosophical discussion. (Remember the Heraclitus quote: "No person ever steps in the same river twice, for it's not the same river and he/she is not the same person.") Here, I just want to avoid keeping two copies of a zillion variables. It is just a convenience to have two instances of a class instead...